elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/AsyncExecutor.scala

```diff
@@ -0,0 +1,47 @@
+package com.sksamuel.elastic4s.http
+
+import cats.Functor
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+
+/**
+  *  A typeclass that wraps an asynchronous computation.
+  *
+  *  Ideally, we'd paramterize on Functor instead of extending, but extending allows us to introduce this
+  *  without breaking API changes. TODO: In the next major release, F[_] should be parameterized on Functor.
+  */
```

Comment:
Can you explain what you mean by F should be parameterized on Functor? F <: Functor or something else ?

Reply:
I mean `[F[_]: Functor]` (or `implicit functor: Functor[F]`). Basically, cats (and cats-effect) provide typeclass instances of Functor for IO and Scala Future, so it's sort of redundant for us to provide our own instance. However, using them requires importing the appropriate typeclass implicits at the callsite, meaning users of elastic4s would need to update their code to add the imports. 

Requiring that AsyncExecutor implement its own Functor definition allows us to avoid this. However, it means we're doing extra work and have slightly more code to maintain. 

Honestly though, what I would do is remove the dependency on cats in core: It really ties us into an ecosystem that we're not leveraging beyond `Show`, which is something we can implement just fine without cats. Plus it makes providing a scalaz effect a little weird: scalaz users would be required to have a dependency on both cats and scalaz.

If we opted to pull cats out of core, I would define a `Functor` typeclass within elastic4s-http and then provide instances for it from cats/scalaz `Functor` in the cats/scalaz packages. Then we'd be fully agnostic to either ecosystem. 

Just my two cents.

Reply:
I like your thinking, and I'm going to do just that.
Your changes are excellent, and they will get merged into master (with some changes to incorporate the other open PR). It'll be a week or two though as I've got other things to tend to at the moment, plus this kind of big change I want to get right and will be in 7.0.0 only not a 6.x release - but we can push out a 7.0.0-pre-pre-alpha release which is just 6.x.x in all but name.

Reply:
Sounds great. Let me know if I can do anything to help.

Once this lands in master I'll make a PR with the cats-effect package I have sitting in another branch, and potentially look into the scalaz integration.

Reply:
I did a cats effect, monix and scalaz implementation in my pulsar client.
https://github.com/sksamuel/pulsar4s
See what you think

It was easier to do over there as I had a few hours, this elastic change is going to take a lot longer as I want to change every request implicit.